src/validataclass/validators/allow_empty_string.py

```diff
@@ -0,0 +1,80 @@
+Use of this source code is governed by an MIT-style license that can be found in the LICENSE file.
+"""
+
+from copy import deepcopy
+from typing import Any, Optional
+
+from validataclass.exceptions import InvalidTypeError
+from .validator import Validator
+
+__all__ = [
+    'AllowEmptyString',
+]
+
+
+class AllowEmptyString(Validator):
+    """
+    Special validator that wraps another validator, but allows empty string ('') as input value.
+
+    By default, the wrapper returns empty string ('') for empty string ('') as input value. Optionally a default value
+    can be specified in the constructor that will be returned instead of empty string ('').
+
+    If the wrapped validator raises an `InvalidTypeError`, `AllowEmptyString` will add empty string ('')
+    to its 'expected_types' parameter and reraise it.
+
+    Examples:
+
+    ```
+    # Accepts strings and empty string (''), returns both unmodified (e.g. "foo" -> "foo", "" -> "")
+    AllowEmptyString(StringValidator())
+
+    # Accepts strings and empty string (''), returns a default string for empty string ('')
+     (e.g. '' -> "no value given!", but "foo" -> "foo")
+    AllowEmptyString(StringValidator(), default='no value given!')
+    ```
+
+    Valid input: `` or any data accepted by the wrapped validator
+    Output: `` (or default value specified in constructor) or the output of the wrapped validator
+    """
+
+    # Default value returned in case the input is empty string ('')
+    default_value: Any
+
+    # Validator used in case the input is not empty string ('')
+    wrapped_validator: Validator
+
+    def __init__(self, validator: Validator, *, default: Any = ''):
+        """
+        Create a AllowEmptyString wrapper validator.
+
+        Parameters:
+            validator: Validator that will be wrapped (required)
+            default: Value of any type that is returned instead of empty string ('') (default: ``)
+        """
+        # Check parameter validity
+        if not isinstance(validator, Validator):
+            raise TypeError('AllowEmptyString requires a Validator instance.')
+
+        self.wrapped_validator = validator
+        self.default_value = default
+
+    def validate(self, input_data: Any, **kwargs) -> Optional[Any]:
+        """
+        Validate input data.
+
+        If the input is empty string (''), return empty string ('') (or the value specified in the `default` parameter).
+        Otherwise, pass the input to the wrapped validator and return its result.
+        """
+        if input_data == "":
+            return deepcopy(self.default_value)
+
+        try:
+            # Call wrapped validator for all values other than empty string ('')
+            return self.wrapped_validator.validate_with_context(input_data, **kwargs)
+        except InvalidTypeError as error:
+            # If wrapped validator raises an InvalidTypeError, add '' to its 'expected_types' list and reraise it
+            error.add_expected_type(str(''))
```

Comment:
The type of an empty string still is `str`, so we should add the type `str` to the list of expected types, not the empty string itself.

Currently, there is a small problem in the `InvalidTypeError` exception, though: It doesn't check for duplicates, so if the wrapped validator accepts strings, the exception will already contain `str`, leading to the output `expected_types: ['str', 'str']` which of course would be a bit silly. ^^

We can easily solve this, though, by adding a simple `if not in expected_types` to the `InvalidTypeError.add_expected_type()` method (make sure to do the `_type_to_string` conversion before the check).
(This should also get a small unit test in `InvalidTypeErrorTest`.)

Reply:
Changed to **str**. And also change to not add duplicates made in **InvalidTypeError.add_expected_type()** + test